# Preprocessing — PropNavigator

**Input:** `data/fe/featured_properties.csv` (39,621 × 52, from feature engineering)
**Output:** a clean, model-ready table (`data/pp/preprocessed_properties.csv`).

**Scope:** *clean → impute → remove outliers → shape features → drop/reorder columns.*
**Not here:** encoding & scaling — those are fit on the **training split** inside the model
pipeline (doing them here would leak test info into the encoders).

This notebook mirrors `src/data_preprocessing/pre_process_data.py` step for step. Each section
says **what** and **why**, with a **before/after** so every decision is visible.

## Setup & load

In [1]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)

df = pd.read_csv('../../data/fe/featured_properties.csv')
print('shape:', df.shape)
df.head(3)

shape: (39621, 52)


,property_name,link,society,price_in_cr,price_per_sqft,areaWithType,carpetArea,bedRoom,bathroom,balcony,address,floorNum,facing,overlooking,agePossession,cornerProperty,furnishing,parking,nearbyLocations,description,features,latitude,longitude,property_id,property_type,sector,floornum,total_floor,area,age_possession_category,is_corner,covered_parking,open_parking,total_parking,has_ac,has_power_backup,has_gym,has_pool,has_club,has_lift,has_servant_room,is_gated,ov_park,ov_pool,ov_club,ov_main_road,ov_sea,ov_others,dist_to_cyber_city,dist_to_golf_road,dist_to_airport,dist_to_manesar
0,"2 BHK Flat in Sector 89, Gurgaon",https://www.99acres.com/2-bhk-bedroom-apartmen...,smart world gems,1.4,11706.0,1196 sqft Superbuiltup Area,NaN,2,2,2.0,"Sector 89, Gurgaon",1 of 4 Floors,east,Others,0 to 1 Year Old,NaN,semi-furnished,"{""C"":1}","['Dwarka Expressway', 'Genesis Hospital Sector...",The flat is north-Facing. Constructed on a sup...,"['Wheel Chair Friendly', 'Full Power Backup', ...",28.421078,76.949658,W89454606,Flat,sector 89,1.0,4.0,1196.0,New Property,0,1.0,0.0,1.0,0,1,1,1,1,0,0,1,0,0,0,0,0,1,17.442210,17.194023,22.471484,7.577067
1,"3 BHK Builder Floor in Suncity, Gurgaon",https://www.99acres.com/3-bhk-bedroom-independ...,NaN,2.5,11111.0,2250 sqft Super Area,NaN,3,3,2.0,"Suncity, Gurgaon",G of 3 Floors,east,Park/Garden,10+ Year Old,NaN,unfurnished,"{""O"":2}","['Sector 54 chowk metro station', 'Sector metr...",Excellent opportunity to buy a 250 sqyds groun...,"['Vaastu Compliant', 'Visitor Parking Availabl...",28.435375,77.115020,L90358202,Independent Builder Floor,suncity,0.0,3.0,2250.0,Old Property,0,0.0,2.0,2.0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,7.326951,1.580030,12.263103,21.726818
2,"3 BHK Flat in Sector 92, Gurgaon",https://www.99acres.com/3-bhk-bedroom-apartmen...,signature global city 92,1.3,11918.0,1091 sqft Superbuiltup Area,NaN,3,2,3.0,"Sector 92, Gurgaon",1 of 4 Floors,north-west,"Park/Garden, Club, Others, Sea facing, Main Ro...",0 to 1 Year Old,Y,unfurnished,"[""N""]","['Iris Broadway Mall', 'Dwarka Expressway', 'I...",The flat is facing the north-West direction. T...,"['Wheel Chair Friendly', 'Full Power Backup', ...",28.412378,76.916649,L89188331,Flat,sector 92,1.0,4.0,1091.0,New Property,1,0.0,0.0,0.0,1,1,1,0,0,0,0,0,1,1,1,1,1,1,21.144571,20.962383,25.996431,6.862454


## Diagnosis — what needs fixing?

The checklist the rest of the notebook works through.

In [2]:
nulls = df.isna().sum()
print('missing values:')
print(nulls[nulls > 0].sort_values(ascending=False))
print()
print('bathroom == 0 :', int((df['bathroom'] == 0).sum()))
print('age Undefined :', int((df['age_possession_category'] == 'Undefined').sum()))
print('facing values :', df['facing'].unique().tolist())

missing values:
cornerProperty        16732
carpetArea            16726
overlooking            5500
society                4367
total_floor            2027
floorNum               2020
floornum               2020
features                844
price_per_sqft          489
total_parking           364
parking                 364
covered_parking         364
open_parking            364
nearbyLocations         259
dist_to_cyber_city      187
longitude               187
latitude                187
dist_to_airport         187
dist_to_golf_road       187
dist_to_manesar         187
balcony                  23
dtype: int64

bathroom == 0 : 14
age Undefined : 150
facing values : ['east', 'north-west', 'not available', 'north', 'north-east', 'south-east', 'south', 'south-west', 'west']


## Step 1 — Fix `bathroom == 0`
A home can't have zero bathrooms — fall back to the bedroom count.

In [3]:
mask = df['bathroom'] == 0
df.loc[mask, 'bathroom'] = df.loc[mask, 'bedRoom']
print('bathroom == 0 fixed on', int(mask.sum()), 'rows')

bathroom == 0 fixed on 14 rows


## Step 2 — Fill "absence-means-zero" columns

Separate fills, one concept each (matching the `.py` functions):
- **floors** — missing only for independent houses (no floor number) → 0
- **parking** — no parking info recorded → 0
- **balcony** — missing → 0

In [4]:
df[['floornum', 'total_floor']] = df[['floornum', 'total_floor']].fillna(0)     # floors
df[['covered_parking', 'open_parking', 'total_parking']] = \
    df[['covered_parking', 'open_parking', 'total_parking']].fillna(0)          # parking
df['balcony'] = df['balcony'].fillna(0)                                          # balcony
print('floor/parking/balcony nulls filled with 0')

floor/parking/balcony nulls filled with 0


## Step 3 — Impute missing distances with the median
The four `dist_to_*` features are missing for 187 rows with un-fixable coordinates. Fill each
with its median rather than dropping real listings over a coordinate glitch.

In [5]:
dist_cols = ['dist_to_cyber_city', 'dist_to_golf_road', 'dist_to_airport', 'dist_to_manesar']
for c in dist_cols:
    df[c] = df[c].fillna(df[c].median())
print('distance nulls filled:', int(df[dist_cols].isna().sum().sum()))

distance nulls filled: 0


## Step 4 — Clean `facing`
"not available" / missing isn't really missing — for a buyer, *not knowing* the facing is its
own state. Make it an explicit **`unknown`** category.

In [6]:
df['facing'] = df['facing'].fillna('unknown').replace(
    {'not available': 'unknown', 'na': 'unknown', '': 'unknown'})
print('facing:', df['facing'].unique().tolist())

facing: ['east', 'north-west', 'unknown', 'north', 'north-east', 'south-east', 'south', 'south-west', 'west']


## Step 5 — Clean `furnishing`
Same idea, its own step.

In [7]:
df['furnishing'] = df['furnishing'].fillna('unknown').replace(
    {'not available': 'unknown', 'na': 'unknown', '': 'unknown'})
print('furnishing:', df['furnishing'].unique().tolist())

furnishing: ['semi-furnished', 'unfurnished', 'furnished', 'unknown']


## Step 6 — Impute `Undefined` age (3-pass mode, readable)

Borrow the most common age from similar listings, in progressively broader groups:
1. same **sector + property type**, 2. same **sector**, 3. same **property type**.

Row-based and explicit (the same code that's in the `.py`).

In [8]:
def mode_based_imputation(row, df):
    if row['age_possession_category'] == 'Undefined':
        mv = df[(df['sector'] == row['sector']) &
                (df['property_type'] == row['property_type'])]['age_possession_category'].mode()
        return mv.iloc[0] if not mv.empty else row['age_possession_category']
    return row['age_possession_category']

def mode_based_imputation2(row, df):
    if row['age_possession_category'] == 'Undefined':
        mv = df[df['sector'] == row['sector']]['age_possession_category'].mode()
        return mv.iloc[0] if not mv.empty else row['age_possession_category']
    return row['age_possession_category']

def mode_based_imputation3(row, df):
    if row['age_possession_category'] == 'Undefined':
        mv = df[df['property_type'] == row['property_type']]['age_possession_category'].mode()
        return mv.iloc[0] if not mv.empty else row['age_possession_category']
    return row['age_possession_category']

print('before — Undefined:', int((df['age_possession_category'] == 'Undefined').sum()))
df['age_possession_category'] = df.apply(lambda r: mode_based_imputation(r, df), axis=1)
df['age_possession_category'] = df.apply(lambda r: mode_based_imputation2(r, df), axis=1)
df['age_possession_category'] = df.apply(lambda r: mode_based_imputation3(r, df), axis=1)
print('after  — Undefined:', int((df['age_possession_category'] == 'Undefined').sum()))

before — Undefined: 150
after  — Undefined: 0


## Step 7 — Group rare `society` values into `other`
~1,860 societies, ~4,400 blank, many appearing once or twice. Fill blanks with `other` and fold
societies appearing < 5 times into `other` (society mainly serves the recommender, not the price model).

In [9]:
df['society'] = df['society'].fillna('other')
counts = df['society'].value_counts()
rare = counts[counts < 5].index
before = df['society'].nunique()
df['society'] = df['society'].where(~df['society'].isin(rare), 'other')
print(f'society cardinality: {before} -> {df["society"].nunique()}')

society cardinality: 1862 -> 663


## Step 8 — Outliers: **investigate first, then remove only errors**

We do **not** clip by a percentile. A ₹150 Cr penthouse is real; a ₹1 Cr flat of 12 sqft is a
typo. So first we *look*, then remove only genuine errors — keeping real luxury **and** real
affordable homes. A handy tell: the **implied price-per-sqft** = `price ÷ area`.

### 8.1 — Investigate the extremes

In [10]:
df['implied_ppsf'] = (df['price_in_cr'] * 1e7 / df['area']).round(0)
look = ['property_type', 'sector', 'society', 'price_in_cr', 'area', 'bedRoom', 'implied_ppsf']

print('===== TOP 10 by PRICE =====')
print(df.nlargest(10, 'price_in_cr')[look].to_string())
print()
print('===== BOTTOM 8 by PRICE =====')
print(df.nsmallest(8, 'price_in_cr')[look].to_string())

===== TOP 10 by PRICE =====
      property_type           sector            society  price_in_cr       area  bedRoom  implied_ppsf
4966           Flat        sector 95              other       1000.0  1701718.0        4        5876.0
17700          Flat        sector 42  dlf the camellias        150.0    15162.0        4       98932.0
3776           Flat        sector 42  dlf the camellias        135.0     9500.0        5      142105.0
7058           Flat        sector 42  dlf the camellias        130.0    13095.0        4       99275.0
15906          Flat        sector 42  dlf the camellias        130.0    15162.0        5       85741.0
20425          Flat        sector 42  dlf the camellias        130.0    15162.0        5       85741.0
1106           Flat  dlf golf course  dlf the camellias        125.0     9500.0        5      131579.0
23155          Flat        sector 42  dlf the camellias        125.0    13095.0        5       95456.0
19937          Flat        sector 42  dlf the

In [11]:
print('===== TOP 10 by AREA =====')
print(df.nlargest(10, 'area')[look].to_string())
print()
print('===== BOTTOM 8 by AREA =====')
print(df.nsmallest(8, 'area')[look].to_string())

===== TOP 10 by AREA =====
                   property_type               sector                                       society  price_in_cr       area  bedRoom  implied_ppsf
18926                       Flat                sohna                                         other         0.75  4312440.0        3           2.0
4966                        Flat            sector 95                                         other      1000.00  1701718.0        4        5876.0
12373          Independent House                other                                         other         2.80   623999.0        3          45.0
27921  Independent Builder Floor      sohna sector 33  central park flower valley aqua front towers         2.55   450000.0        3          57.0
27222          Independent House      galleria market                                         other        15.00    99000.0       10        1515.0
5191           Independent House            sector 70                              bptp vis

> **Genuine vs noise:**
>
> | Extreme | Example | Verdict |
> |---|---|---|
> | High price + big area | Camellias ₹150 Cr, 15,000 sqft | ✅ genuine luxury |
> | Low price + small area | EWS flat ₹18 L, 365 sqft | ✅ genuine affordable |
> | Huge price + impossible area | ₹1000 Cr, **1.7M sqft** | ❌ noise (project area) |
> | Tiny area | ₹1 L, **12 sqft** | ❌ noise |
> | Big "builder floor" | 45,000 sqft single floor | ❌ noise |
> | Contradictory | 60 sqft @ ₹4.75L/sqft; 4.3M sqft @ ₹2/sqft | ❌ noise |
>
> **Noise = impossible area OR price/area that contradict — not high price.** So three checks:

### 8.2 — Univariate: area must be physically possible (per type)
Flats & floors are single units (can't be huge); independent houses sit on plots. The flat cap sits
just above the genuine Camellias ceiling (15,162 sqft).

In [12]:
AREA_CAP = {'Flat': 16000, 'Independent Builder Floor': 10000, 'Independent House': 30000}
before = len(df)
cap = df['property_type'].map(AREA_CAP)
df = df[(df['area'] >= 180) & (df['area'] <= cap)].reset_index(drop=True)
print(f'area outliers removed: {before - len(df)}')

area outliers removed: 193


### 8.3 — Bivariate (price vs area): implied ₹/sqft must be sane
Below ~1,500 ⇒ area overstated; above ~2.5 L ⇒ area understated.

In [13]:
before = len(df)
implied = df['price_in_cr'] * 1e7 / df['area']
df = df[implied.between(1500, 250000)].reset_index(drop=True)
print(f'price-vs-area outliers removed: {before - len(df)}')

price-vs-area outliers removed: 48


### 8.4 — Bivariate (area vs bedrooms): space per bedroom
Under **150 sqft/bedroom** is not a real layout (catches "540 sqft / 13 bedrooms"). We use 150 not
200 because for independent houses `area` is often the plot *footprint* of a multi-floor house, so
a genuine 540 sqft plot with G+2 floors legitimately has 3 bedrooms (ratio 180).

In [14]:
before = len(df)
df = df[(df['area'] / df['bedRoom']) >= 150].reset_index(drop=True)
print(f'area-per-bedroom outliers removed: {before - len(df)}')
print('max bedrooms now:', int(df['bedRoom'].max()))

area-per-bedroom outliers removed: 279
max bedrooms now: 30


## Step 9 — Categorize `floornum`
Instead of keeping the raw floor number, bucket it into Low-/Mid-/High-rise (the raw column is
dropped in Step 10).

In [15]:
def _cat(floor):
    if pd.isna(floor):
        return 'Undefined'
    f = float(floor)
    if f <= 5:
        return 'Low-rise'
    elif f <= 15:
        return 'Mid-rise'
    return 'High-rise'

df['floornum_category'] = df['floornum'].apply(_cat)
print(df['floornum_category'].value_counts().to_dict())

{'Low-rise': 19656, 'Mid-rise': 14437, 'High-rise': 5008}


## Step 10 — Drop unused columns, then drop duplicates

- **IDs & text**, **raw pre-engineered** columns, **`price_per_sqft`** (leakage), **raw coords**
  (replaced by distances), and **raw `floornum`** (replaced by the category).
- Then drop **exact duplicates** — once the IDs are gone, a few rows that appeared in more than one
  source file become identical.

In [16]:
drop_cols = [
    'property_name', 'link', 'areaWithType', 'carpetArea', 'address', 'floorNum',
    'overlooking', 'agePossession', 'cornerProperty', 'parking', 'nearbyLocations',
    'description', 'features', 'property_id',
    'price_per_sqft', 'latitude', 'longitude', 'floornum', 'implied_ppsf',
]
df = df.drop(columns=[c for c in drop_cols if c in df.columns], errors='ignore')

before = len(df)
df = df.drop_duplicates().reset_index(drop=True)
print(f'dropped {before - len(df)} duplicate rows; {df.shape[1]} columns remain')

dropped 36 duplicate rows; 35 columns remain


## Step 11 — Reorder columns into a clean structure

In [17]:
desired_order = [
    'property_type', 'society', 'sector', 'price_in_cr', 'area',
    'bedRoom', 'bathroom', 'balcony', 'floornum_category', 'total_floor',
    'facing', 'furnishing', 'age_possession_category', 'is_corner',
    'covered_parking', 'open_parking', 'total_parking',
    'has_ac', 'has_power_backup', 'has_gym', 'has_pool', 'has_club',
    'has_lift', 'has_servant_room', 'is_gated',
    'ov_park', 'ov_pool', 'ov_club', 'ov_main_road', 'ov_sea', 'ov_others',
    'dist_to_cyber_city', 'dist_to_golf_road', 'dist_to_airport', 'dist_to_manesar',
]
df = df[[c for c in desired_order if c in df.columns]]
list(df.columns)

['property_type',
 'society',
 'sector',
 'price_in_cr',
 'area',
 'bedRoom',
 'bathroom',
 'balcony',
 'floornum_category',
 'total_floor',
 'facing',
 'furnishing',
 'age_possession_category',
 'is_corner',
 'covered_parking',
 'open_parking',
 'total_parking',
 'has_ac',
 'has_power_backup',
 'has_gym',
 'has_pool',
 'has_club',
 'has_lift',
 'has_servant_room',
 'is_gated',
 'ov_park',
 'ov_pool',
 'ov_club',
 'ov_main_road',
 'ov_sea',
 'ov_others',
 'dist_to_cyber_city',
 'dist_to_golf_road',
 'dist_to_airport',
 'dist_to_manesar']

## Final check

In [18]:
print('final shape :', df.shape)
print('total nulls :', int(df.isna().sum().sum()))
print('Undefined age:', int((df['age_possession_category'] == 'Undefined').sum()))
df.head(3)

final shape : (39065, 35)
total nulls : 0
Undefined age: 0


,property_type,society,sector,price_in_cr,area,bedRoom,bathroom,balcony,floornum_category,total_floor,facing,furnishing,age_possession_category,is_corner,covered_parking,open_parking,total_parking,has_ac,has_power_backup,has_gym,has_pool,has_club,has_lift,has_servant_room,is_gated,ov_park,ov_pool,ov_club,ov_main_road,ov_sea,ov_others,dist_to_cyber_city,dist_to_golf_road,dist_to_airport,dist_to_manesar
0,Flat,smart world gems,sector 89,1.4,1196.0,2,2,2.0,Low-rise,4.0,east,semi-furnished,New Property,0,1.0,0.0,1.0,0,1,1,1,1,0,0,1,0,0,0,0,0,1,17.442210,17.194023,22.471484,7.577067
1,Independent Builder Floor,other,suncity,2.5,2250.0,3,3,2.0,Low-rise,3.0,east,unfurnished,Old Property,0,0.0,2.0,2.0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,7.326951,1.580030,12.263103,21.726818
2,Flat,signature global city 92,sector 92,1.3,1091.0,3,2,3.0,Low-rise,4.0,north-west,unfurnished,New Property,1,0.0,0.0,0.0,1,1,1,0,0,0,0,0,1,1,1,1,1,1,21.144571,20.962383,25.996431,6.862454


> **Result:** ~39.1k × 35, zero nulls, no `Undefined` ages, luxury retained. Matches
> `pre_process_data.py` exactly. Ready for the model pipeline (encoding + scaling on the train split).

**Save:**

In [ ]:
# df.to_csv('../../data/pp/preprocessed_properties.csv', index=False)
# print('saved:', df.shape)